src/adapters/ip_camera_adapter/mod.rs

```diff
@@ -0,0 +1,304 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use controller::Controller;
+use iron::{ Request, Response, IronResult };
+use iron::headers::{ ContentType, AccessControlAllowOrigin };
+use iron::status::Status;
+use router::Router;
+use service::{ Service, ServiceAdapter, ServiceProperties };
+use std::fs;
+use std::io::{ BufWriter, Error, ErrorKind };
+use std::io::prelude::*;
+use std::path::Path;
+use std::thread;
+use uuid::Uuid;
+
+extern crate serde;
+extern crate serde_json;
+
+extern crate time;
+
+extern crate hyper;
+use self::hyper::header::{ Authorization, Basic, Connection };
+
+// TODO: The camera username and password need to be persisted per-camera
+static CAMERA_USERNAME: &'static str = "admin";
+static CAMERA_PASSWORD: &'static str = "password";
+static SNAPSHOT_DIR: &'static str = "snapshots";
+
+pub fn get_bytes(url: String) -> Result<Vec<u8>, Error> {
+    let client = hyper::Client::new();
+    let get_result = client.get(&url)
+                           .header(
+                               Authorization(
+                                   Basic {
+                                       username: CAMERA_USERNAME.to_owned(),
+                                       password: Some(CAMERA_PASSWORD.to_owned())
+                                   }
+                               )
+                           )
+                           .header(Connection::close())
+                           .send();
+    let mut res = match get_result {
+        Ok(res) => res,
+        Err(err) => {
+            return Err(Error::new(ErrorKind::Other,
+                                  format!("GET on {} failed: {}", url, err)));
+        }
+    };
+
+    if res.status != Status::Ok {
+        debug!("ip_camera_adapter: get_bytes {}", res.status);
+        return Err(Error::new(ErrorKind::Other,
+                              format!("GET on {} failed: {}", url, res.status)));
+    }
+    
+    let mut image = Vec::new();
+    match res.read_to_end(&mut image) {
+        Ok(_) => Ok(image),
+        Err(err) => Err(Error::new(ErrorKind::Other,
+                                   format!("read of image data from {} failed: {}", url, err)))
+    }
+}
+
+struct IpCameraService<T> {
+    controller: T,
+    properties: ServiceProperties,
+    ip: String,
+    snapshot_dir: String,
+}
+
+impl<T: Controller> IpCameraService<T> {
+    fn new(controller: T, ip: &str) -> Self {
+        debug!("Creating IpCameraService");
+        let service_id = Uuid::new_v4().to_simple_string();
```

Comment:
We need stable ID to associate a piece of HW with persisted meta-data. For IP devices, the MAC address could do.